From e694b49de7a7deee8a84d205b11389242ddf9def Mon Sep 17 00:00:00 2001
From: Jason Rogena <j.rogena@cgiar.org>
Date: Wed, 28 May 2014 22:17:14 +0300
Subject: [PATCH 176/213] Version 0.9.7: Extension Personnel and locale during
 registration

Extension personnel changed from edittext to Spinner in farmer
registration

New Spinner for setting preferred locale also added in farmer
registration
---
 MistroFarmer/src/main/AndroidManifest.xml          |   4 +-
 .../mistro/farmer/FarmerRegistrationActivity.java  | 128 +++++++++++++++++++--
 .../ilri/mistro/farmer/backend/DataHandler.java    |   1 +
 .../cgiar/ilri/mistro/farmer/backend/Locale.java   |  36 ++++++
 .../cgiar/ilri/mistro/farmer/carrier/Farmer.java   |  13 +++
 .../main/res/layout/activity_cow_registration.xml  |   2 +-
 .../res/layout/activity_farmer_registration.xml    |  21 +++-
 MistroFarmer/src/main/res/values/strings_en.xml    |   1 +
 8 files changed, 192 insertions(+), 14 deletions(-)

diff --git a/MistroFarmer/src/main/AndroidManifest.xml b/MistroFarmer/src/main/AndroidManifest.xml
index 95d499e..375fae8 100644
--- a/MistroFarmer/src/main/AndroidManifest.xml
+++ b/MistroFarmer/src/main/AndroidManifest.xml
@@ -1,8 +1,8 @@
 <?xml version="1.0" encoding="utf-8"?>
 <manifest xmlns:android="http://schemas.android.com/apk/res/android"
     package="org.cgiar.ilri.mistro.farmer"
-    android:versionCode="6"
-    android:versionName="0.9.6" >
+    android:versionCode="7"
+    android:versionName="0.9.7" >
 
     <uses-sdk
         android:minSdkVersion="7"
diff --git a/MistroFarmer/src/main/java/org/cgiar/ilri/mistro/farmer/FarmerRegistrationActivity.java b/MistroFarmer/src/main/java/org/cgiar/ilri/mistro/farmer/FarmerRegistrationActivity.java
index b663fce..19786c0 100644
--- a/MistroFarmer/src/main/java/org/cgiar/ilri/mistro/farmer/FarmerRegistrationActivity.java
+++ b/MistroFarmer/src/main/java/org/cgiar/ilri/mistro/farmer/FarmerRegistrationActivity.java
@@ -16,8 +16,10 @@ import android.provider.Settings;
 import android.telephony.TelephonyManager;
 import android.util.Log;
 import android.view.View;
+import android.widget.ArrayAdapter;
 import android.widget.Button;
 import android.widget.EditText;
+import android.widget.Spinner;
 import android.widget.TextView;
 import android.widget.Toast;
 
@@ -29,8 +31,12 @@ import com.actionbarsherlock.view.MenuItem;
 import org.cgiar.ilri.mistro.farmer.backend.DataHandler;
 import org.cgiar.ilri.mistro.farmer.backend.Locale;
 import org.cgiar.ilri.mistro.farmer.carrier.Farmer;
+import org.json.JSONArray;
 import org.json.JSONObject;
 
+import java.util.ArrayList;
+import java.util.List;
+
 public class FarmerRegistrationActivity extends SherlockActivity implements MistroActivity, View.OnClickListener,LocationListener
 {
     public static final String TAG="FarmerRegistrationActivity";
@@ -41,8 +47,10 @@ public class FarmerRegistrationActivity extends SherlockActivity implements Mist
     private String longitude;
     private TextView fullNameTV;
     private EditText fullNameET;
+    private TextView preferredLanguageTV;
+    private Spinner preferredLanguageS;
     private TextView extensionPersonnelTV;
-    private EditText extensionPersonnelET;
+    private Spinner extensionPersonnelS;
     private TextView mobileNumberTV;
     private EditText mobileNumberET;
     private TextView numberOfCowsTV;
@@ -58,7 +66,10 @@ public class FarmerRegistrationActivity extends SherlockActivity implements Mist
     private String mobileNoETEmptyWarning;
     private  LocationManager locationManager;
     private String loadingPleaseWait;
+    private List<String> vetNames;
     private boolean isInFarm;
+    private List<String> languages;
+    private int preferredLanguageIndex;
 
     private Farmer farmer;
     @Override
@@ -69,13 +80,17 @@ public class FarmerRegistrationActivity extends SherlockActivity implements Mist
         //DataHandler.requestPermissionToUseSMS(this);
 
         //init child views
+        languages = Locale.getAllLanguages(this);
         cacheData = true;
         isInFarm = false;
+        preferredLanguageIndex = -1;
 
         fullNameTV=(TextView)this.findViewById(R.id.full_name_tv);
         fullNameET=(EditText)this.findViewById(R.id.full_name_et);
+        preferredLanguageTV = (TextView)this.findViewById(R.id.preferred_language_tv);
+        preferredLanguageS = (Spinner)this.findViewById(R.id.preferred_language_s);
         extensionPersonnelTV=(TextView)this.findViewById(R.id.extension_personnel_tv);
-        extensionPersonnelET=(EditText)this.findViewById(R.id.extension_personnel_et);
+        extensionPersonnelS=(Spinner)this.findViewById(R.id.extension_personnel_s);
         mobileNumberTV=(TextView)this.findViewById(R.id.mobile_number_tv);
         mobileNumberET=(EditText)this.findViewById(R.id.mobile_number_et);
         TelephonyManager telephonyManager=(TelephonyManager)this.getSystemService(Context.TELEPHONY_SERVICE);
@@ -88,6 +103,9 @@ public class FarmerRegistrationActivity extends SherlockActivity implements Mist
 
         //init text according to locale
         initTextInViews();
+
+        FetchVetsThread fetchVetsThread = new FetchVetsThread();
+        fetchVetsThread.execute(0);
     }
 
     public boolean onCreateOptionsMenu(Menu menu) {
@@ -107,20 +125,17 @@ public class FarmerRegistrationActivity extends SherlockActivity implements Mist
     private void cacheEditTextData(){
         if(cacheData) {
             DataHandler.setSharedPreference(this, DataHandler.SP_KEY_FRA_FULL_NAME, fullNameET.getText().toString());
-            DataHandler.setSharedPreference(this, DataHandler.SP_KEY_FRA_EXTENSION_PERSONNEL, extensionPersonnelET.getText().toString());
             DataHandler.setSharedPreference(this, DataHandler.SP_KEY_FRA_MOBILE_NUMBER, mobileNumberET.getText().toString());
         }
     }
 
     private void restoreEditTextData(){
         fullNameET.setText(DataHandler.getSharedPreference(this, DataHandler.SP_KEY_FRA_FULL_NAME, ""));
-        extensionPersonnelET.setText(DataHandler.getSharedPreference(this, DataHandler.SP_KEY_FRA_EXTENSION_PERSONNEL, ""));
         mobileNumberET.setText(DataHandler.getSharedPreference(this, DataHandler.SP_KEY_FRA_MOBILE_NUMBER, ""));
     }
 
     private void clearEditTextDataCache(){
         DataHandler.setSharedPreference(this, DataHandler.SP_KEY_FRA_FULL_NAME, "");
-        DataHandler.setSharedPreference(this, DataHandler.SP_KEY_FRA_EXTENSION_PERSONNEL, "");
         DataHandler.setSharedPreference(this, DataHandler.SP_KEY_FRA_MOBILE_NUMBER, "");
 
         cacheData = false;
@@ -142,7 +157,30 @@ public class FarmerRegistrationActivity extends SherlockActivity implements Mist
         {
             fullNameET.setText(farmer.getFullName());
             Log.d(TAG, "Full name: " + farmer.getFullName());
-            extensionPersonnelET.setText(farmer.getExtensionPersonnel());
+            List<String> tmpVetNames = new ArrayList<String>();
+            tmpVetNames.add("");
+            String selectedEP = farmer.getExtensionPersonnel();
+            int epSelection = 0;
+            if(selectedEP!= null && selectedEP.length() > 0){
+                tmpVetNames.add(selectedEP);
+                epSelection = 1;
+            }
+            ArrayAdapter<String> epArrayAdapter = new ArrayAdapter<String>(this,android.R.layout.simple_spinner_item,tmpVetNames);
+            epArrayAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
+            extensionPersonnelS.setAdapter(epArrayAdapter);
+            extensionPersonnelS.setSelection(epSelection);
+
+            String prefLocale = farmer.getPreferredLocale();
+            if(prefLocale != null && prefLocale.length() > 0){
+                for(int i = 0; i < languages.size(); i++){
+                    if(languages.get(i).equals(Locale.getLanguage(this, prefLocale))){
+                        preferredLanguageIndex = i;
+                    }
+                }
+            }
+
+            Log.d(TAG, "Preferred locale: "+farmer.getPreferredLocale());
+
             Log.d(TAG,"Extension Personnel: "+farmer.getExtensionPersonnel());
             mobileNumberET.setText(farmer.getMobileNumber());
             Log.d(TAG,"Mobile number: "+farmer.getMobileNumber());
@@ -182,6 +220,14 @@ public class FarmerRegistrationActivity extends SherlockActivity implements Mist
         extensionPersonnelTV.setText(Locale.getStringInLocale("extension_p", this));
         mobileNumberTV.setText(" * " + Locale.getStringInLocale("mobile_number", this));
         numberOfCowsTV.setText(" * "+Locale.getStringInLocale("number_of_cows",this));
+        preferredLanguageTV.setText(" * "+Locale.getStringInLocale("preferred_language", this));
+
+        ArrayAdapter<String> languageAdapter = new ArrayAdapter<String>(this,android.R.layout.simple_spinner_item, languages);
+        languageAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
+        preferredLanguageS.setAdapter(languageAdapter);
+
+        if(preferredLanguageIndex != -1) preferredLanguageS.setSelection(preferredLanguageIndex);
+
         registerButton.setText(Locale.getStringInLocale("register",this));
         gpsAlertDialogTitle=Locale.getStringInLocale("enable_gps",this);
         gpsAlertDialogText=Locale.getStringInLocale("reason_for_enabling_gps",this);
@@ -205,7 +251,17 @@ public class FarmerRegistrationActivity extends SherlockActivity implements Mist
             farmer=new Farmer();
         }
         farmer.setFullName(fullNameET.getText().toString());
-        farmer.setExtensionPersonnel(extensionPersonnelET.getText().toString());
+        if(vetNames != null && extensionPersonnelS.getSelectedItemPosition() != -1){
+            farmer.setExtensionPersonnel(vetNames.get(extensionPersonnelS.getSelectedItemPosition()));
+        }
+        else{
+            farmer.setExtensionPersonnel("");
+        }
+
+        if(languages != null && preferredLanguageS.getSelectedItemPosition() != -1){
+            farmer.setPreferredLocale(Locale.getLocaleCode(this, languages.get(preferredLanguageS.getSelectedItemPosition())));
+        }
+
         farmer.setMobileNumber(mobileNumberET.getText().toString());
         farmer.setCowNumber((numberOfCowsET.getText().toString()==null||numberOfCowsET.getText().toString().length()==0) ? 0:Integer.parseInt(numberOfCowsET.getText().toString()));//Integer.parseInt(numberOfCowsET.getText().toString())
         if(isInFarm) {
@@ -441,4 +497,62 @@ public class FarmerRegistrationActivity extends SherlockActivity implements Mist
         }
     }
 
+    private class FetchVetsThread extends AsyncTask<Integer, Integer, String>{
+        private ProgressDialog progressDialog;
+
+        @Override
+        protected void onPreExecute() {
+            super.onPreExecute();
+            progressDialog = ProgressDialog.show(FarmerRegistrationActivity.this, "",loadingPleaseWait, true);
+        }
+
+        @Override
+        protected String doInBackground(Integer... params) {
+            Log.d(TAG, "Fetching vets from server");
+
+            return DataHandler.sendDataToServer(FarmerRegistrationActivity.this, "", DataHandler.FARMER_FETCH_VETS_URL, true);
+        }
+
+        @Override
+        protected void onPostExecute(String result) {
+            super.onPostExecute(result);
+            progressDialog.dismiss();
+
+            if(result == null){
+                Toast.makeText(FarmerRegistrationActivity.this, Locale.getStringInLocale("problem_connecting_to_server", FarmerRegistrationActivity.this), Toast.LENGTH_LONG).show();
+            }
+            else{
+                try{
+                    Log.d(TAG, "result is "+result);
+                    JSONArray vetJsonArray = new JSONArray(result);
+                    vetNames = new ArrayList<String>();
+                    vetNames.add("");
+                    for(int i = 0; i < vetJsonArray.length(); i++){
+                        vetNames.add(vetJsonArray.getJSONObject(i).getString("name"));
+                    }
+
+                    int selection = 0;
+                    if(farmer != null){
+                        String selectedEP = farmer.getExtensionPersonnel();
+                        if(selectedEP != null && selectedEP.length() > 0){
+                            for(int i =0; i < vetNames.size(); i++){
+                                if(vetNames.get(i).equals(selectedEP)){
+                                    selection = i;
+                                }
+                            }
+                        }
+                    }
+
+                    ArrayAdapter<String> epArrayAdapter = new ArrayAdapter<String>(FarmerRegistrationActivity.this, android.R.layout.simple_spinner_item, vetNames);
+                    epArrayAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
+                    extensionPersonnelS.setAdapter(epArrayAdapter);
+                    extensionPersonnelS.setSelection(selection);
+                }
+                catch (Exception e){
+                    e.printStackTrace();
+                }
+            }
+        }
+    }
+
 }
diff --git a/MistroFarmer/src/main/java/org/cgiar/ilri/mistro/farmer/backend/DataHandler.java b/MistroFarmer/src/main/java/org/cgiar/ilri/mistro/farmer/backend/DataHandler.java
index 9ddd827..1df1e3f 100644
--- a/MistroFarmer/src/main/java/org/cgiar/ilri/mistro/farmer/backend/DataHandler.java
+++ b/MistroFarmer/src/main/java/org/cgiar/ilri/mistro/farmer/backend/DataHandler.java
@@ -74,6 +74,7 @@ public class DataHandler
     //private static final String BASE_URL="http://192.168.14.102/~jason/ngombe_planner/WebServer";
     //private static final String BASE_URL="http://172.26.23.48/~jason/ngombe_planner/WebServer";
     public static final String FARMER_REGISTRATION_URL="/php/farmer/registration.php";
+    public static final String FARMER_FETCH_VETS_URL="/php/farmer/fetch_vets.php";
     public static final String FARMER_AUTHENTICATION_URL="/php/farmer/authentication.php";
     public static final String FARMER_SIM_CARD_REGISTRATION_URL="/php/farmer/sim_card_registration.php";
     public static final String FARMER_FETCH_COW_IDENTIFIERS_URL="/php/farmer/fetch_cow_identifiers.php";
diff --git a/MistroFarmer/src/main/java/org/cgiar/ilri/mistro/farmer/backend/Locale.java b/MistroFarmer/src/main/java/org/cgiar/ilri/mistro/farmer/backend/Locale.java
index 0e21a23..0d16b7e 100644
--- a/MistroFarmer/src/main/java/org/cgiar/ilri/mistro/farmer/backend/Locale.java
+++ b/MistroFarmer/src/main/java/org/cgiar/ilri/mistro/farmer/backend/Locale.java
@@ -5,6 +5,8 @@ import android.content.SharedPreferences;
 import android.util.Log;
 import org.cgiar.ilri.mistro.farmer.R;
 import java.lang.reflect.Field;
+import java.util.ArrayList;
+import java.util.List;
 
 /**
  * Created by jason on 9/5/13.
@@ -179,4 +181,38 @@ public class Locale {
         return sharedPreferences.getString(SHARED_PREFERENCES_KEY, LOCALE_ENGLISH);*/
         return DataHandler.getSharedPreference(context, DataHandler.SP_KEY_LOCALE, LOCALE_ENGLISH);
     }
+
+    public static String getLocaleCode(Context context, String language){
+        String code = "";
+        if(language.equals(context.getString(R.string.english))) code = LOCALE_ENGLISH;
+        else if(language.equals(context.getString(R.string.swahili))) code = LOCALE_SWAHILI;
+        else if(language.equals(context.getString(R.string.kalenjin))) code = LOCALE_KALENJIN;
+        else if(language.equals(context.getString(R.string.kikabrasi))) code = LOCALE_KIKABRAS;
+        else if(language.equals(context.getString(R.string.luhya))) code = LOCALE_LUHYA;
+
+        return code;
+    }
+
+    public static String getLanguage(Context context, String localeCode){
+        String language = "";
+
+        if(localeCode.equals(LOCALE_ENGLISH)) language = context.getString(R.string.english);
+        if(localeCode.equals(LOCALE_SWAHILI)) language = context.getString(R.string.swahili);
+        if(localeCode.equals(LOCALE_KALENJIN)) language = context.getString(R.string.kalenjin);
+        if(localeCode.equals(LOCALE_KIKABRAS)) language = context.getString(R.string.kikabrasi);
+        if(localeCode.equals(LOCALE_LUHYA)) language = context.getString(R.string.luhya);
+
+        return language;
+    }
+
+    public static List<String> getAllLanguages(Context context){
+        List<String> languages = new ArrayList<String>();
+        languages.add(context.getString(R.string.english));
+        languages.add(context.getString(R.string.swahili));
+        languages.add(context.getString(R.string.luhya));
+        languages.add(context.getString(R.string.kikabrasi));
+        languages.add(context.getString(R.string.kalenjin));
+
+        return languages;
+    }
 }
diff --git a/MistroFarmer/src/main/java/org/cgiar/ilri/mistro/farmer/carrier/Farmer.java b/MistroFarmer/src/main/java/org/cgiar/ilri/mistro/farmer/carrier/Farmer.java
index fbdfaf3..e1d3fc9 100644
--- a/MistroFarmer/src/main/java/org/cgiar/ilri/mistro/farmer/carrier/Farmer.java
+++ b/MistroFarmer/src/main/java/org/cgiar/ilri/mistro/farmer/carrier/Farmer.java
@@ -29,6 +29,7 @@ public class Farmer implements Parcelable, Serializable
     private String latitude;
     private String simCardSN;
     private String mode;
+    private String preferredLocale;
     private boolean isInFarm;
 
     public Farmer()
@@ -41,6 +42,7 @@ public class Farmer implements Parcelable, Serializable
         latitude="";
         simCardSN ="";
         mode = "";
+        preferredLocale = "";
         isInFarm = false;
     }
 
@@ -103,6 +105,14 @@ public class Farmer implements Parcelable, Serializable
         }
     }
 
+    public String getPreferredLocale() {
+        return preferredLocale;
+    }
+
+    public void setPreferredLocale(String preferredLocale) {
+        this.preferredLocale = preferredLocale;
+    }
+
     public void addCow(Cow cow)
     {
         this.cows.add(cow);
@@ -194,6 +204,7 @@ public class Farmer implements Parcelable, Serializable
         dest.writeString(mode);
         if(isInFarm) dest.writeInt(1);
         else dest.writeInt(0);
+        dest.writeString(preferredLocale);
     }
 
     public void readFromParcel(Parcel in)
@@ -210,6 +221,7 @@ public class Farmer implements Parcelable, Serializable
         int isInFarm = in.readInt();
         if(isInFarm == 1) this.isInFarm = true;
         else this.isInFarm = false;
+        this.preferredLocale = in.readString();
     }
 
     public static final Creator<Farmer> CREATOR=new Creator<Farmer>()
@@ -245,6 +257,7 @@ public class Farmer implements Parcelable, Serializable
             jsonObject.put("latitude",((latitude==null) ? "":latitude));
             jsonObject.put("simCardSN",((simCardSN ==null) ? "": simCardSN));
             jsonObject.put("mode",((mode ==null) ? "": mode));
+            jsonObject.put("preferredLocale", preferredLocale);
         }
         catch (JSONException e)
         {
diff --git a/MistroFarmer/src/main/res/layout/activity_cow_registration.xml b/MistroFarmer/src/main/res/layout/activity_cow_registration.xml
index 5d98f35..82af928 100644
--- a/MistroFarmer/src/main/res/layout/activity_cow_registration.xml
+++ b/MistroFarmer/src/main/res/layout/activity_cow_registration.xml
@@ -232,7 +232,7 @@
             android:layout_marginLeft="@dimen/activity_f_r_edit_text_margin_left"
             android:textSize="@dimen/normal_text_size"
             android:textColor="@color/text_input_color"
-            android:inputType="text"/>
+            android:inputType="textCapWords|textPersonName"/>
         <TextView
             android:id="@+id/straw_number_tv"
             android:layout_height="wrap_content"
diff --git a/MistroFarmer/src/main/res/layout/activity_farmer_registration.xml b/MistroFarmer/src/main/res/layout/activity_farmer_registration.xml
index 395c06b..515503d 100644
--- a/MistroFarmer/src/main/res/layout/activity_farmer_registration.xml
+++ b/MistroFarmer/src/main/res/layout/activity_farmer_registration.xml
@@ -43,20 +43,33 @@
                 android:textColor="@color/text_input_color"
                 android:textSize="@dimen/normal_text_size"/>
         <TextView
+            android:id="@+id/preferred_language_tv"
+            android:layout_width="wrap_content"
+            android:layout_height="wrap_content"
+            android:layout_marginTop="@dimen/activity_f_r_text_view_margin_top"
+            android:textSize="@dimen/normal_text_size"/>
+        <Spinner
+            android:id="@+id/preferred_language_s"
+            android:layout_width="match_parent"
+            android:layout_height="wrap_content"
+            android:layout_marginTop="@dimen/activity_f_r_edit_text_margin_top"
+            android:layout_marginLeft="@dimen/activity_f_r_edit_text_margin_left"
+            android:textSize="@dimen/normal_text_size"
+            android:textColor="@color/text_input_color" />
+        <TextView
             android:id="@+id/extension_personnel_tv"
             android:layout_width="wrap_content"
             android:layout_height="wrap_content"
             android:layout_marginTop="@dimen/activity_f_r_text_view_margin_top"
             android:textSize="@dimen/normal_text_size"/>
-        <EditText
-            android:id="@+id/extension_personnel_et"
+        <Spinner
+            android:id="@+id/extension_personnel_s"
             android:layout_width="match_parent"
             android:layout_height="wrap_content"
             android:layout_marginTop="@dimen/activity_f_r_edit_text_margin_top"
             android:layout_marginLeft="@dimen/activity_f_r_edit_text_margin_left"
             android:textSize="@dimen/normal_text_size"
-            android:textColor="@color/text_input_color"
-            android:inputType="textCapWords|textPersonName"/>
+            android:textColor="@color/text_input_color" />
         <TextView
             android:id="@+id/number_of_cows_tv"
             android:layout_width="wrap_content"
diff --git a/MistroFarmer/src/main/res/values/strings_en.xml b/MistroFarmer/src/main/res/values/strings_en.xml
index 6dcb6ae..1b27239 100644
--- a/MistroFarmer/src/main/res/values/strings_en.xml
+++ b/MistroFarmer/src/main/res/values/strings_en.xml
@@ -282,4 +282,5 @@
     <string name="enter_sire_etn_en">Enter the sire\'s ear tag number</string>
     <string name="sire_owner_en">Sire\'s Owner</string>
     <string name="name_sire_owner_en">Name of Sire\'s Owner</string>
+    <string name="preferred_language_en">Preferred language</string>
 </resources>
-- 
2.7.1

